wagtail/admin/views/generic/lock.py

```diff
@@ -0,0 +1,87 @@
+from django.urls import reverse
+from django.utils import timezone
+from django.utils.text import capfirst
+from django.utils.translation import gettext as _
+from django.views import View
+
+from wagtail.admin import messages
+from wagtail.admin.utils import get_valid_next_url_from_request
+from wagtail.log_actions import log
+from wagtail.models import DraftStateMixin
+
+
+class LockView(View):
+    model = None
+    pk_url_kwarg = "pk"
+    success_url_name = None
+    success_message_extra_tags = "lock"
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.next_url = get_valid_next_url_from_request(request)
+        self.object = self.get_object()
+
+    def get_success_url(self):
+        if not self.success_url_name:
+            raise ImproperlyConfigured(
+                "Subclasses of wagtail.admin.views.generic.lock.LockView must provide a "
+                "success_url_name attribute or a get_success_url method"
+            )
+        if self.next_url:
+            return self.next_url
+        return reverse(self.success_url_name, args=[quote(self.object.pk)])
+
+    def get_success_message(self):
+        return None
+
+    def post(self, request, *args, **kwargs):
+        self.perform_action()
+        success_message = self.get_success_message()
+        if success_message:
+            messages.success(
+                request,
+                success_message,
+                extra_tags=self.success_message_extra_tags,
+            )
+        return redirect(self.get_success_url())
+
+    def perform_action(self):
+        if self.object.locked:
+            return
+        self.object.locked = True
+        self.object.locked_by = self.request.user
+        self.object.locked_at = timezone.now()
+        self.object.save(update_fields=["locked", "locked_by", "locked_at"])
+        log(instance=self.object, action="wagtail.lock", user=self.request.user)
+
+    def get_object(self):
+        pk = self.kwargs[self.pk_url_kwarg]
+        if isinstance(pk, str):
+            pk = unquote(pk)
+        return get_object_or_404(self.model, pk=pk)
+
+
+class UnlockView(LockView):
+    success_message_extra_tags = "unlock"
+
+    def perform_action(self):
+        if not self.object.locked:
+            return
+        self.object.locked = False
+        self.object.locked_by = None
+        self.object.locked_at = None
+        self.object.save(update_fields=["locked", "locked_by", "locked_at"])
+        log(instance=self.object, action="wagtail.unlock", user=self.request.user)
+
+    def get_success_message(self):
+        title = str(self.object)
+        if isinstance(self.object, DraftStateMixin) and self.object.latest_revision:
+            title = self.object.latest_revision.object_str
+
+        return _("%(model_name)s '%(title)s' is now unlocked.") % {
+            "model_name": capfirst(self.model._meta.verbose_name),
```

Comment:
Should we add the model name information like this, or remove it and just use the title? The `capfirst` might be a bit awkward if it's in the middle of the sentence in other languages, but I think Django admin does this, too. I also don't know if this translates well for languages with gendered nouns.

Reply:
It'd be nice to include the model name - looks like we do for success messages on other snippet views. Maybe the `capfirst` could be moved to outside the `_("...")`?